### DISCUSSION TITLE
how to provide additional field (material, auxvariable or function) being read from csv-files, so that another matprop can depend on this field

### DISCUSSION THREAD

**Discussion by arndflatten** [idaholab/moose#26294] [Q&A General] [answered]
Dear Moose users,

I am posting this question in general section, as this is a conceptual question, despite I want to apply this for a PorousFlow problem:

I want to make certain material properties like permeability and porosity depend on an additional „field“, here damage. That damage is provided at elements via csv (at elCentroids for now, but elQP data could be provided as well). I understand that I will subclass from one of the existing PorousFlowPermeabilityXXXX classes, but

1.	I am struggling with the moose concept, whether that damage field should be another material property, an AuxVariable, or a PiecewiseLinearFunction? Does it make any difference, or is one way better than the other?

Now, reading this element data from csv, I think the userobject-class ReadPropertyFile will be appropriate, but I expect the damage field to change each timestep (remaining constant in time for that timestep, but changes at begin of next timestep). So i am providing many (dozens of) csv files, like damage0001.csv, damage0002.csv, etc. 

2.	do i need to declare as many userobject-objects of type=ReadPropertyFile, one for each csv? Then, one might use the control logic to enable the appropriate reader-object for a certain timestep, but how does the damage field, that my material property depends on, link to get data from the activated reader-object but not from the others? I see a potential shared interest as discussed in https://github.com/idaholab/moose/discussions/25953, but my problem is still a bit different.

Many thanks for helping me find the right direction,

**Comment by GiudGiud** [comment, marked as answer]
hello

1)
Yes there is a difference between the three. 
a) As a material property:
- the additional field will be stored with a negligible memory cost, as material properties are computed on a per-element basis for every quadrature point.
- the matprop will be constantly updated, on linear iterations
- you can propagate derivatives with regards to the variable when using automatic differentiation. This is important if you are solving for this extra damage field as well.
b) As an auxvariable
- the field will be computed once using an initial condition or an auxkernel, then only updated with auxkernels if needed on the execute_on specified in the input
- the quadrature point values are obtained by evaluating the shape functions for the finite element family you use for that variable
- derivatives are lost if you are trying to use AD
c) as a function
- the field must only depend on space and time. Any other use (depend on variables) is a little hacky
- the function is evaluated on-the-fly at quadrature points, so the information is always up-to-date, even though it's mostly constant for a time step

Since you are planning to use a UserObject to read the data from a CSV, I believe you can use:
- https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html
and maybe there is a material property in PorousFlow that already does it.

Unfortunately, yes for 2. PropertyReadFile should and will one day support multiple CSV files for each time step. But for now the only thing implemented is a single CSV file. Then turning on/off functions with controls will not work, because functions are referenced by their name and the name is unique. 
However you can turn on/off with controls both auxkernels and materials (less sure for that second one but please let me know if you try it)

So if you used an auxkernel with let's say 10 CSV files for ten time steps you need:
- 10 PropertyReadFile to read each file
- 10 functions PiecewiseConstrantFromCSV, one for each UO
- 10 auxkernels FunctionAux that set the SAME auxvariable `damage` using a different function
- 10 Controls objects that turn on/off the auxkernels on each time step

I agree this is not idea. It's either that or implementing support for multiple files in PropertyReadFile. We have a feature request iirc, just no one to work on it

Guillaume

**Reply by arndflatten** [reply]
Thank you Guillaume, for your detailed answer!

Just one follow-up question on this: Do you know, if all the PropertyReadFile objects (as 10 UOs for 10 csv-files) do read these csv files all at the beginning of the analysis, or, more ideal, read them just when the control enables the specific UO? 

The latter would be ideal to both to (i) reduce time to setup before calculation starts and potentially reduce memory needed to store just one csv-field instead of all csv-fields, as well as to (ii) have the option to get the csv-files provided by an external process while the moose job is already running. Thanks!

Arnd

**Reply by GiudGiud** [reply]
They read them at construction time, not at activation time.
See the call to `  readData();` in the constructor.

I can see the new for both (i) and (ii). I dont think it would be too much work to move this call to initialSetup() instead or even later if you want to try

**Comment by GiudGiud** [comment]
@arndflatten I realized building 10 functions was not how we wanted to do this in MOOSE so I gave a try a specifying multiple CSV files to a single reader, and one is read every execution of the object (either every time step or using Controls)

see this PR https://github.com/idaholab/moose/pull/26395 if you want to try it out

**Reply by arndflatten** [reply]
Thank you for your work on this, @GiudGiud. I will try this..

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Thank you for making this, this is such an awesome feature to have.

I just have a quick question:

if we list a multiple csv files to prop_file_name, for example `prop_file_name = 'damage1.csv, damage2.csv, damage3.csv'`
how do we control which file do we want to read with respect to time? for example If I want damage1 at t = 1, damage2 at t = 2, damage3 at t = 3?

with my limited knowledge of [Controls], I thought it only can control a whole `PropertyReadFile` block :

I see you have `_prop_file_names[_current_file_index]` 
and `_current_file_index(declareRestartableData<unsigned int>("file_index", 0)),`
is there a way that I can control `_current_file_index` from input file?

I couldn't find `file_index` from `PropertyReadFile::validParams()`, did you forget or it's somewhere else

Below is my current workflow

```
[UserObjects]

 [damage_read_Damage_1003]
  type = PropertyReadFile
  prop_file_name = 'Damage/Damage_1003.csv'
  nprop = 1
  read_type = element
  execute_on = '  TIMESTEP_BEGIN' 
 []
 [damage_read_Damage_1004]
  type = PropertyReadFile
  prop_file_name = 'Damage/Damage_1004.csv'
  nprop = 1
  read_type = element
  execute_on = '  TIMESTEP_BEGIN' 
 []
 [damage_read_Damage_1005]
  type = PropertyReadFile
  prop_file_name = 'Damage/Damage_1005.csv'
  nprop = 1
  read_type = element
  execute_on = '  TIMESTEP_BEGIN' 
 []

[]

then use the Materials to contain these data from csv files

[damage_Damage_1003]
  type = GenericConstantArrayBEH
  prop_name = damage
  read_prop_user_object = damage_read_Damage_1003
 []
 [damage_Damage_1004]
  type = GenericConstantArrayBEH
  prop_name = damage
  read_prop_user_object = damage_read_Damage_1004
 []
 [damage_Damage_1005]
  type = GenericConstantArrayBEH
  prop_name = damage
  read_prop_user_object = damage_read_Damage_1005
 []

then pass it to permability calculartion

 [permeability]
  type = PorousFlowPermeabilityConstCSV
  read_prop_user_object = perm_read
  damage = damage
 []

and below is the [Controls]

[Controls]
 [controls_Damage_1003] 
  type = ConditionalFunctionEnableControl 
  enable_objects = 'Materials::damage_Damage_1003 UserObjects::damage_read_Damage_1003'
  conditional_function = 'if(t>= 0 & t< 2,1,0)' 
  execute_on = ' INITIAL TIMESTEP_BEGIN TIMESTEP_END' 
 [] 
 [controls_Damage_1004] 
  type = ConditionalFunctionEnableControl 
  enable_objects = 'Materials::damage_Damage_1004 UserObjects::damage_read_Damage_1004'
  conditional_function = 'if(t>= 2& t< 3,1,0)' 
  execute_on = ' INITIAL TIMESTEP_BEGIN TIMESTEP_END' 
 [] 
 [controls_Damage_1005] 
  type = ConditionalFunctionEnableControl 
  enable_objects = 'Materials::damage_Damage_1005 UserObjects::damage_read_Damage_1005'
  conditional_function = 'if(t>= 3 & t< 4,1,0)' 
  execute_on = ' INITIAL TIMESTEP_BEGIN TIMESTEP_END' 
 [] 
```

as you can see each ConditionalFunctionEnableControl block is used to control each UserObjects::damage_read_Damage_XXXX

Thank you @GiudGiud

**Reply by GiudGiud** [reply]
`_current_file_index` isnt controllable from the input file. 
I m not sure we can do controllable and restartable at the same time. I think we can. INL is off right now I ll ask when people come back.

With what I implemented, I would:
- run the user object reader with execute_on on timestep_end
- make the controls turn on the user object for exactly one time step at the desired times. 
How big are your timesteps? Do you have more than one between t=1, t=2 ?

btw should be
prop_file_name = 'damage1.csv damage2.csv damage3.csv'

**Reply by Traiwit** [reply]
Happy new year @GiudGiud 

> make the controls turn on the user object for exactly one time step at the desired times.
> How big are your timesteps? Do you have more than one between t=1, t=2 ?

it really depends on physical time (for transient), sometimes 0.5 years or months

I still dont understand how do I hold the data from each .csv file and change it over time if we cannot control them?

for example if we read all damage files from a single `PropertyReadFile` userobj 

the input file would be

```
 [damage_read_Damage_files]
  type = PropertyReadFile
  prop_file_name = 'Damage/Damage_1001.csv Damage/Damage_1002.csv Damage/Damage_1003.csv' 
  nprop = 1
  read_type = element
  execute_on = 'timestep_end' 
 []
```

then lets say we have something like (below) to hold the read data

```
[damage_Damage_1003]
  type = GenericConstantArrayBEH
  prop_name = damage
  read_prop_user_object = damage_read_Damage_files  (it will remain like this for all steps since we only have 1 PropertyReadFile obj ?)
 []
```

for the above example, what do we control if we want to update the damage for each timestep?

Maybe i'm just confused, could you please explain more.

Thank you!

Kind regards,
Trai

**Reply by GiudGiud** [reply]
If you change each time step, you do not need to use Controls.

Use Controls only if you want to skip time steps.
Btw capability got merged in MOOSE like two days ago

**Reply by GiudGiud** [reply]
Happy New Year to you too!

**Reply by Traiwit** [reply]
> If you change each time step, you do not need to use Controls.
> 
> Use Controls only if you want to skip time steps. Btw capability got merged in MOOSE like two days ago

are you saying, it will automatically assign each file to each timestep in sequence?

for example below is my actual `Executioner`, I have 15 timesteps there from `TimeStepper`
do you mean if I have files: damage1 to 15 under `PropertyReadFile`, it will automatically assign if simulation go in sequence smoothly (no non-converged step and cut back time -- adding extra steps to the simulation)

damage1 at t=0.500 
damage2 at t=0.500 
...
damage14 at t = 3.250 
damage15 at t=3.333

Kind regards,
Trai

```
[Executioner]
  type = Transient   # always use 'Transient' (not 'Steady'), see [Kernels]
  solve_type = NEWTON
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre    boomeramg'
  nl_rel_tol = 1e-8
  nl_abs_tol = 10000.0
  nl_max_its = 50
  l_tol = 1e-05
  l_max_its = 100
  # line_search = 'bt'
  [TimeStepper]
    type = TimeSequenceStepper
    time_sequence = '0.500 1.000 1.500 2.417 2.500 2.583 2.667 2.750 2.833 2.917 3.000 3.083 3.167 3.250 3.333'
  []
  end_time = 3.333
[]
```

**Reply by GiudGiud** [reply]
It should. If it does not please let me know I ll need to get back on it

>  it will automatically assign if simulation go in sequence smoothly (no non-converged step and cut back time -- adding extra steps to the simulation)

yup that s the big issue with executing on time steps. If you fail one the reading will go out of sync for sure.

**Reply by Traiwit** [reply]
Hi @GiudGiud 

Awesome! it works as you said, but just letting you know this thing always shows up at the end of step 1, which shouldn't be the case

`
*** Info ***
Last file specified has been read. The file will no longer be updated.
`

Kind regards,
Trai

![image](https://github.com/idaholab/moose/assets/79562439/6562f7bd-a833-4943-ac4f-fcad35cea218)

**Reply by GiudGiud** [reply]
ah this is coming from the other ones (poro_read, ...)

Thanks for letting me know. I ll silence the log when there's only 1 file specified.
